### PR TITLE
 Make option API fail when using a disallowed option name in non-produciton environments

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

pdibGW-1Yx-p2#comment-1399

This PR updates the option API to make it return a `WP_Error` when using a disallowed option name in non-produciton environments. This is to force developers to update the option permissions when adding new options to prevent https://github.com/woocommerce/woocommerce/issues/32863 from happing again.

I also updated the option allowlist for WC Test helper options and new add product pages.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set your `WP_ENVIRONMENT_TYPES` in wp-config to a non-production environment.
2. Make a request to GET `/wp-json/wc-admin/options?options=woocommerce_shop_page_id`
3. The response should be 403 with a  message `Sorry, you cannot view these options, please remember to update the option permissions in Options API to allow viewing these options in non-production environments.`

### CODE REVIEW

**Review by rjchow — DISMISSED** [review]
Tests well! Confirmed that it works on JN but not on localhost dev environment, without any fiddling with wp-config.php

Note to any other testers, the easiest way to test this is probably to copy the request for an existing options request by filtering for 'options' in the browser devtools, and then changing the options name in the command.

File: plugins/woocommerce/src/Admin/API/Options.php [hunk 1]
```diff
@@ -99,6 +107,11 @@ public function user_has_permission( $option, $request, $is_update = false ) {
 			return $permissions[ $option ];
 		}
 
+		// Don't allow to update options in non-production environments if the option is not whitelisted. This is to force developers to update the option permissions when adding new options.
+		if ( 'production' !== wp_get_environment_type() ) {
+			return false;
+		}
+
 		wc_deprecated_function( 'Automattic\WooCommerce\Admin\API\Options::' . ( $is_update ? 'update_options' : 'get_options' ), '3.1' );
```
Comment (ilyasfoo) [inline plugins/woocommerce/src/Admin/API/Options.php]: I wonder if it's worth updating the version reference `3.1` here? It used to refer to WooCommerce Admin version 3.1. I believe the closest version the deployed admin code is paired with [WooCommerce 6.3](https://developer.woocommerce.com/2022/03/08/woocommerce-6-3-released/)

**Review by ilyasfoo — DISMISSED** [review]
Tested well! 🚀

File: plugins/woocommerce/src/Admin/API/Options.php (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce/src/Admin/API/Options.php]: Yeah, good catch. Updated in 2ca9b8d15f3e6ec027de632a2132f1204bdd6409.

Could you please approve this PR again? Thank you! 😄

**Review by ilyasfoo — APPROVED** [review]
Sorry for the delay, LGTM! 🚢

**Pull request merged**